### DISCUSSION TITLE
Get the exit code from a command without capturing?

### DISCUSSION THREAD

**Discussion by halloleo** [xonsh/xonsh#5286] [Q&A] [answered]
I have a long running progress-lines-outputting command. In subproc mode I run it like this:

    my_command arg1 arg2

Now I have a xonsh script doing exactly that and in the script, after the command ran I  would like to know, what the exit code of  `my_command` was. How can I find out? `!()` is no good because it captures the output as well; instead I would like the output from the command coming into the terminal character by character while `my_command` is running. I only need the exit code.

So what can I do to get only the exit code in my xonsh script?

Thanks for pointers!

**Comment by halloleo** [comment]
The best I could come up with is:

```xsh
$RAISE_SUBPROC_ERROR = True
import subprocess

try:
    my_command arg1 arg2
except subprocess.CalledProcessError as e:
    # do things in case of failure

# normal continuation
```

This doesn't give the the exit code _number_ but at least a mechanism to see whether exit code was `==0` or not.

**Comment by gforsyth** [comment, marked as answer]
You can use `![]` for this.  That creates a subproc pipeline that doesn't capture output, but has all of the other attributes of `!()`.

```xonsh
ls
# a  b
a = ![ls .]  # note that output is not captured and streams to stdout
# a  b
a.returncode
# 0
a.output  # no output captured
# ''
```

**Reply by anki-code** [reply]
Thanks!

@halloleo feel free to create PR with updating [docs](https://github.com/xonsh/xonsh/tree/main/docs).

Good example:
```xsh
if r := ![ls NONONO]:
    print(f'It works! Return code: {r.returncode}')
else:
    print(f'It fails! Return code: {r.returncode}')

# ls: cannot access 'NONONO': No such file or directory
# It fails! Return code: 2
```

**Reply by halloleo** [reply]
Thanks for this. I will try it out.
